Implement file source to integrate Galaxy with RSpace

domgz opened the pull request:
This PR implements a FilesSource that interacts with an [RSpace](https://www.researchspace.com/) instance and includes a file source template so that users can define their own file sources based on it.

RSpace is an electronic lab notebook (ELN) designed for documenting, managing and sharing scientific data. The central concepts in RSpace are the Workspace, the Inventory and the Gallery.

The [Workspace](https://documentation.researchspace.com/article/bzgr8ea9e3-the-workspace)'s role is organizing and managing research activities. It allows creating documents, folders and research notebooks. The [Inventory](https://documentation.researchspace.com/category/zpizk20kgx-inventory) is a tool to keep track of laboratory resources, such as reagents, samples and equipment. The [Gallery](https://documentation.researchspace.com/article/sl6mo1i9do-the-gallery) is where any research data represented as files is stored.

The FilesSource implemented by this module accesses the Gallery of an RSpace instance. The Gallery is organized in categories (e.g. Images, Audio, Documents, Chemistry), where each category is itself a tree structure with folders and files just like a regular filesystem, although there is a key difference: categories and their subfolders can only contain files of certain types (e.g. jpeg, bmp, png, etc. for the Images category).

This implementation is based on the [`rspace-client-python` library](https://github.com/rspace-os/rspace-client-python), which provides a [PyFilesystem2](https://docs.pyfilesystem.org/) interface to interact with RSpace's Gallery (available as `GalleryFilesystem` in the `rspace_client.eln.fs` module). It consists just of a minimal wrapper (based on the base `PyFilesystem2FilesSource` provided by Galaxy) around the `GalleryFilesystem` class.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. Run RSpace, for example [using Docker Compose](https://github.com/rspace-os/rspace-docker).
  2. Enable the file source template lib/galaxy/files/templates/examples/production_rspace.yaml on your file_source_templates.yml configuration file.
  3. Log-in with the [default account](https://github.com/rspace-os/rspace-docker?tab=readme-ov-file#installing-rspace-on-docker) "sysadmin1" and generate an API Key (under the My RSpace section).
  4. Create a new RSpace file source under Preferences ->  Manage Your Remote File Sources -> Create -> RSpace, configuring the endpoint and API key.
  5. Import files from the RSpace file source (for example, one of the sample images) to your history.
  6. Export files from your history to RSpace using the ["Export datasets"](http://127.0.0.1:8000/?tool_id=export_remote) tool.
  7. Export your history to the RSpace file source.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

domgz committed:
Implement file source to integrate Galaxy with RSpace

This commit implements a FilesSource that interacts with an RSpace [1] instance and includes a file source template so that users can define their own file sources based on it.

RSpace is an electronic lab notebook
(ELN) designed for documenting, managing and sharing scientific data. The central concepts in RSpace are the Workspace,
the Inventory and the Gallery.

The Workspace's role is organizing and managing research activities. It allows creating documents, folders and research
notebooks [2]. The Inventory is a tool to keep track of laboratory resources, such as reagents, samples and equipment
[3]. The Gallery is where any research data represented as files is stored [4].

The FilesSource implemented by this module accesses the Gallery of an RSpace instance. The Gallery is organized in
categories (e.g. Images, Audio, Documents, Chemistry), where each category is itself a tree structure with folders and
files just like a regular filesystem, although there is a key difference: categories and their subfolders can only
contain files of certain types (e.g. jpeg, bmp, png, etc. for the Images category).

This implementation is based on the rspace-client-python` library [5], which provides a PyFilesystem2 [6] interface
to interact with RSpace's Gallery (available as `GalleryFilesystem` in the rspace_client.eln.fs` module). It
consists just of a minimal wrapper (based on the base `PyFilesystem2FilesSource` provided by Galaxy) around the
`GalleryFilesystem` class.

References:

- [1] https://www.researchspace.com/
- [2] https://documentation.researchspace.com/article/bzgr8ea9e3-the-workspace
- [3] https://documentation.researchspace.com/category/zpizk20kgx-inventory
- [4] https://documentation.researchspace.com/article/sl6mo1i9do-the-gallery
- [5] https://github.com/rspace-os/rspace-client-python
- [6] https://docs.pyfilesystem.org/

bgruening commented on lib/galaxy/config/sample/user_preferences_extra_conf.yml.sample:
```diff
@@ -183,3 +183,17 @@ preferences:
               type: text
               required: True
+
+    # Used in file_sources_conf.yml
```
I guess its fine if we not advertise this option anymore. Instead, we should use the templates. I would remove this addition and the one to the file_sources_conf.yml file

domgz committed:
Remove examples showing how to configure RSpace as global file source

There are two ways to enable file sources: globally or as file source templates. File source templates allow each user to define their own file sources, providing their credentials in the process; the newly defined file sources are only shown in the list of file sources after they have been set up. Global file sources are always shown on the list of file sources; credentials may be provided on the user preferences page.

Everything that is "personal", for example Dropbox or your own S3 buckets should be enabled using file source templates. Everything that is public AND requires no credentials (or equivalently, it is meant to be accessible to all users of the instance, as it may happen in a private lab) should also be enabled using file source templates. The remaining may be configured as global file sources.

RSpace falls in the first category, thus it makes sense to discourage configuring it as a global file source.

davelopez committed:
Fix isort linting

davelopez reviewed (approved):
Thank you @kysrpex!

Looks and works fine in my local testing! :+1:
